File: sdk/typescript/src/namespaces/selection.ts
/**
 * Selection Namespace API
 *
 * Provides endpoints for agent selection including
 * scoring, team composition, and role assignment.
 */

import type { AragoraClient } from '../client';

/** Agent score from the selection process */
export interface SelectionScore {
  agent_id: string;
  agent_name: string;
  score: number;
  factors: Record<string, number>;
}

/** Selected team member */
export interface SelectedTeamMember {
  agent_id: string;
  agent_name: string;
  role: string;
  score: number;
}

/** Selection history entry */
export interface SelectionHistoryItem {
  id: string;
  debate_id: string;
  selected_agents: string[];
  strategy: string;
  scores: SelectionScore[];
  created_at: string;
}

/** Team selection request */
export interface SelectTeamRequest {
  task: string;
  team_size?: number;
  required_roles?: string[];
  domain?: string;
  strategy?: string;
}

/** Team selection result */
export interface SelectTeamResult {
  members: SelectedTeamMember[];
  strategy_used: string;
  selection_time_ms: number;
}

/**
 * Selection namespace for agent team composition.
 *
 * @example
 * ```typescript
 * const team = await client.selection.selectTeam({
 *   task: 'API design review',
 *   team_size: 3,
 * });
 * console.log(team.members.map(m => m.agent_name));
 * ```
 */
export class SelectionNamespace {
  constructor(private client: AragoraClient) {}

  /** Score agents for a given task. */
  async scoreAgents(options: {
    task: string;
    domain?: string;
    agent_ids?: string[];
  }): Promise<SelectionScore[]> {
    const response = await this.client.request<{ scores: SelectionScore[] }>(
      'POST',
      '/api/v1/selection/score',
      { body: options }
    );
    return response.scores;
  }

  /** Select a team for a task. */
  async selectTeam(request: SelectTeamRequest): Promise<SelectTeamResult> {
    return this.client.request<SelectTeamResult>(
      'POST',
      '/api/v1/selection/team',
      { body: request }
    );
  }

  /** Get the best agent for a task. */
  async getBestAgent(options: {
    task: string;
    domain?: string;
  }): Promise<SelectionScore> {
    return this.client.request<SelectionScore>(
      'POST',
      '/api/v1/agent-selection/best',
      { body: options }
    );
  }

  /** Get selection history. */
  async getHistory(options?: {
    limit?: number;
    offset?: number;
  }): Promise<SelectionHistoryItem[]> {
    const response = await this.client.request<{ history: SelectionHistoryItem[] }>(
      'GET',
      '/api/v1/agent-selection/history',
      { params: options }
    );
    return response.history;
  }

  /**
   * Get team selection.
   */
  async teamSelection(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', '/api/v1/team-selection', { params }) as Promise<Record<string, unknown>>;
  }

  // ===========================================================================
  // Plugin Listings
  // ===========================================================================

  /**
   * List all available scorer plugins.
   *
   * @route GET /api/v1/selection/scorers
   *
   * @example
   * ```typescript
   * const scorers = await client.selection.listScorers();
   * console.log(scorers);
   * ```
   */
  async listScorers(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/selection/scorers') as Promise<Record<string, unknown>>;
  }

  /**
   * List all available team selector plugins.
   *
   * @route GET /api/v1/selection/team-selectors
   *
   * @example
   * ```typescript
   * const selectors = await client.selection.listTeamSelectors();
   * console.log(selectors);
   * ```
   */
  async listTeamSelectors(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/selection/team-selectors') as Promise<Record<string, unknown>>;
  }

  /**
   * List all available role assigner plugins.
   *
   * @route GET /api/v1/selection/role-assigners
   *
   * @example
   * ```typescript
   * const assigners = await client.selection.listRoleAssigners();
   * console.log(assigners);
   * ```
   */
  async listRoleAssigners(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/selection/role-assigners') as Promise<Record<string, unknown>>;
  }
}

File: sdk/python/aragora_sdk/namespaces/selection.py
"""
Selection Namespace API

Provides methods for agent selection:
- Selection criteria management
- Agent matching
- Performance-based selection
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class SelectionAPI:
    """Synchronous Selection API."""

    def __init__(self, client: AragoraClient):
        self._client = client

    def get_defaults(self, **kwargs: Any) -> dict[str, Any]:
        """Get default agent selection settings."""
        return self._client.request("POST", "/api/v1/selection/defaults", json=kwargs)

    def get_plugins(self, **kwargs: Any) -> dict[str, Any]:
        """Get available selection plugins."""
        return self._client.request("POST", "/api/v1/selection/plugins", json=kwargs)

    def score(self, **kwargs: Any) -> dict[str, Any]:
        """Score agents for a task."""
        return self._client.request("POST", "/api/v1/selection/score", json=kwargs)

    def select_team(self, **kwargs: Any) -> dict[str, Any]:
        """Select a team of agents for a task."""
        return self._client.request("POST", "/api/v1/selection/team", json=kwargs)

    def list_scorers(self) -> dict[str, Any]:
        """List available scorer plugins."""
        return self._client.request("GET", "/api/v1/selection/scorers/")

    def get_scorer(self, name: str) -> dict[str, Any]:
        """Get details for a specific scorer plugin."""
        return self._client.request("GET", f"/api/v1/selection/scorers/{name}")

    def list_team_selectors(self) -> dict[str, Any]:
        """List available team selector plugins."""
        return self._client.request("GET", "/api/v1/selection/team-selectors/")

    def get_team_selector(self, name: str) -> dict[str, Any]:
        """Get details for a specific team selector plugin."""
        return self._client.request("GET", f"/api/v1/selection/team-selectors/{name}")

    def list_role_assigners(self) -> dict[str, Any]:
        """List available role assigner plugins."""
        return self._client.request("GET", "/api/v1/selection/role-assigners/")

    def get_role_assigner(self, name: str) -> dict[str, Any]:
        """Get details for a specific role assigner plugin."""
        return self._client.request("GET", f"/api/v1/selection/role-assigners/{name}")


class AsyncSelectionAPI:
    """Asynchronous Selection API."""

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def get_defaults(self, **kwargs: Any) -> dict[str, Any]:
        """Get default agent selection settings."""
        return await self._client.request("POST", "/api/v1/selection/defaults", json=kwargs)

    async def get_plugins(self, **kwargs: Any) -> dict[str, Any]:
        """Get available selection plugins."""
        return await self._client.request("POST", "/api/v1/selection/plugins", json=kwargs)

    async def score(self, **kwargs: Any) -> dict[str, Any]:
        """Score agents for a task."""
        return await self._client.request("POST", "/api/v1/selection/score", json=kwargs)

    async def select_team(self, **kwargs: Any) -> dict[str, Any]:
        """Select a team of agents for a task."""
        return await self._client.request("POST", "/api/v1/selection/team", json=kwargs)

    async def list_scorers(self) -> dict[str, Any]:
        """List available scorer plugins."""
        return await self._client.request("GET", "/api/v1/selection/scorers/")

    async def get_scorer(self, name: str) -> dict[str, Any]:
        """Get details for a specific scorer plugin."""
        return await self._client.request("GET", f"/api/v1/selection/scorers/{name}")

    async def list_team_selectors(self) -> dict[str, Any]:
        """List available team selector plugins."""
        return await self._client.request("GET", "/api/v1/selection/team-selectors/")

    async def get_team_selector(self, name: str) -> dict[str, Any]:
        """Get details for a specific team selector plugin."""
        return await self._client.request("GET", f"/api/v1/selection/team-selectors/{name}")

    async def list_role_assigners(self) -> dict[str, Any]:
        """List available role assigner plugins."""
        return await self._client.request("GET", "/api/v1/selection/role-assigners/")

    async def get_role_assigner(self, name: str) -> dict[str, Any]:
        """Get details for a specific role assigner plugin."""
        return await self._client.request("GET", f"/api/v1/selection/role-assigners/{name}")
